### PR TITLE
feat: app ratings endpoint

### PR DESCRIPTION
**Pull request by matthew-hagemann**
New app ratings endpoint and its tests

### CODE REVIEW

File: proto/ratings_features_app.proto [hunk 1]
```diff
@@ -3,15 +3,28 @@ syntax = "proto3";
...
+}
+
+message Rating {
+  string snap_id = 1;
+  uint64 total_votes = 2;
+  RatingsBand ratings_band = 3;
+}
+
+enum RatingsBand {
+  VeryGood = 0;
```
Comment (tim-hm) [inline proto/ratings_features_app.proto]:
```suggestion
  VERY_GOOD = 0;
```

File: src/features/app/entities.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        return RatingsBand::InsufficientVotes;
+    }
+    let positive_ratings = votes
+        .into_iter()
+        .filter(|vote| vote.vote_up)
+        .collect::<Vec<Vote>>()
+        .len();
+    let adjusted_ratio = confidence_interval_lower_bound(positive_ratings, total_ratings);
+
+    if adjusted_ratio > 0.8 {
```
Comment (tim-hm) [inline src/features/app/entities.rs]: What do you think about moving these intervals into `impl RatingsBand` and adding a from method? For example:

```
impl RatingsBand {
    const VERY_GOOD_UPPER: f64 = 1.0;
    // ...
    const VERY_POOR_LOWER: f64 = 0.0;

    pub fn from_value(value: f64) -> RatingsBand { ... }
}
```

File: src/features/app/entities.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    } else if adjusted_ratio <= 0.55 && adjusted_ratio > 0.45 {
+        return RatingsBand::Neutral;
+    } else if adjusted_ratio <= 0.45 && adjusted_ratio > 0.2 {
+        return RatingsBand::Poor;
+    } else {
+        return RatingsBand::VeryPoor;
+    }
+}
+
+fn confidence_interval_lower_bound(positve_ratings: usize, total_ratings: usize) -> f64 {
```
Comment (tim-hm) [inline src/features/app/entities.rs]: positve -> positive

Comment (tim-hm) [inline src/features/app/entities.rs]: For this function I'd comments on the approach (ie whats the algo) and perhaps link to a good explanation on the internet.

Also, consider moving some of these values into descriptive variable names

File: src/features/app/entities.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    let z: f64 = 1.96; // hardcoded for a ~95% confidence
+    let n: f64 = total_ratings as f64;
+    let phat = positve_ratings as f64 / n;
+    ((phat + (z * z) / (2.0 * n))
+        - z * f64::sqrt((phat * (1.0 - phat) + ((z * z) / (4.0 * n))) / n))
+        / (1.0 + (z * z) / n)
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (tim-hm) [inline src/features/app/entities.rs:121]: I would also test at the insufficient votes boundary (and perhaps move the insufficient votes quantity into a const)

File: tests/mod.rs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+mod user_tests {
```
Comment (tim-hm) [inline tests/mod.rs:1]: I like this very much better than the approach I had previously!

Although interesting that you don't need a `mod.rs` in each sub folder?

**Review by tim-hm — CHANGES_REQUESTED** [review]
Looking very good! A few changes please:

- Can you run clippy locally and address the issues or disable the lint (but only when there's good reason and if we want to disable a lint then lets discuss 🙏)
- I've added some comments to address inline

Once you've made those changes send it back my way please and I'll have a more thorough tinker with the API.

File: tests/mod.rs (hunk 5 above)
Comment (matthew-hagemann) [reply tests/mod.rs:1]: New to me, but apparently its been here for a while :smile: looks like since Rust 2018 its sufficient to just declare the sub modules in the parent directory. As you say, makes it nice to show everything we have in all the sub directories in one place! We can actually rename it from mod.rs if we want.

https://doc.rust-lang.org/edition-guide/rust-2018/path-changes.html

**Comment by matthew-hagemann** [comment]
> Looking very good! A few changes please:
> 
>     * Can you run clippy locally and address the issues or disable the lint (but only when there's good reason and if we want to disable a lint then lets discuss 🙏)
> 
>     * I've added some comments to address inline
> 
> 
> Once you've made those changes send it back my way please and I'll have a more thorough tinker with the API.

Only Clippy lints left are regarding not having made use of those Config structs in our code yet, but that presumably will change in time?

Let me know if you are happy with how I did the documenting of the Wilson's score interval. Tried to keep it concise and < 80 char per line.

**Comment by tim-hm** [comment]
Excellent work.

**Review by tim-hm — APPROVED** [review]
